### PR TITLE
Login page implemention

### PR DESCRIPTION
**Pull request by Himanshusingh770**

### CODE REVIEW

File: src/Components/Login.jsx [hunk 1]
```diff
@@ -0,0 +1,146 @@
+import React, { useState } from 'react';
```
Comment (PiyushMangla) [inline src/Components/Login.jsx]: These are not components learn file structure first

File: src/Components/Login.jsx [hunk 2]
```diff
@@ -0,0 +1,146 @@
...
+  const [showPassword, setShowPassword] = useState(false);
+  const [showErrorToast, setShowErrorToast] = useState(false);
+
+  const { isLoading, error } = useSelector((state) => state.auth);
+  const dispatch = useDispatch();
+  const navigate = useNavigate();
+
+  const emailRegex = /^[^\s@]+@[^\s@]+\.[^\s@]+$/;
+
+  const validateEmail = () => {
```
Comment (PiyushMangla) [inline src/Components/Login.jsx]: create a function in utils and use it from there

File: src/Components/Login.jsx [hunk 3]
```diff
@@ -0,0 +1,146 @@
...
+    return '';
+  };
+
+  const validatePassword = () => {
+    if (!passwordTouched) return '';
+    if (!password) return 'Password is required';
+    return '';
+  };
+
+  const emailError = validateEmail();
```
Comment (PiyushMangla) [inline src/Components/Login.jsx]: these is wrong

File: src/Components/Login.jsx [hunk 4]
```diff
@@ -0,0 +1,146 @@
...
+  const passwordError = validatePassword();
+
+  const isFormValid = !emailError && !passwordError && emailTouched && passwordTouched;
```
Comment (PiyushMangla) [inline src/Components/Login.jsx]: this too

File: src/Components/Login.jsx [hunk 5]
```diff
@@ -0,0 +1,146 @@
...
+  };
+
+  const validatePassword = () => {
+    if (!passwordTouched) return '';
+    if (!password) return 'Password is required';
+    return '';
+  };
+
+  const emailError = validateEmail();
+  const passwordError = validatePassword();
```
Comment (PiyushMangla) [inline src/Components/Login.jsx]: this too

File: src/Components/Login.jsx [hunk 6]
```diff
@@ -0,0 +1,146 @@
...
+
+  const isFormValid = !emailError && !passwordError && emailTouched && passwordTouched;
+
+  const handleSubmit = (e) => {
+    e.preventDefault();
+    setEmailTouched(true);
```
Comment (PiyushMangla) [inline src/Components/Login.jsx]: what is this email touched

File: src/Components/Login.jsx [hunk 7]
```diff
@@ -0,0 +1,146 @@
...
+
+    if (isFormValid) {
+      dispatch(login({ email: email.trim(), password: password.trim() }))
+        .unwrap()
+        .then(() => navigate('/polls'))
+        .catch(() => setShowErrorToast(true));
+    }
+  };
+
+  const handleEmailChange = (e) => {
```
Comment (PiyushMangla) [inline src/Components/Login.jsx]: use only one function for handle change

File: src/Components/Login.jsx [hunk 8]
```diff
@@ -0,0 +1,146 @@
+import React, { useState } from 'react';
+import { useDispatch, useSelector } from 'react-redux';
+import { login } from '../slices/authSlice';
+import { useNavigate } from 'react-router-dom';
+import { Form, Button, Container, Card, Spinner, ToastContainer, Toast } from 'react-bootstrap';
+import { EyeFill, EyeSlashFill } from 'react-bootstrap-icons';
+
+const Login = () => {
+  const [email, setEmail] = useState('');
+  const [password, setPassword] = useState('');
```
Comment (PiyushMangla) [inline src/Components/Login.jsx]: create an object

File: src/Components/Login.jsx [hunk 9]
```diff
@@ -0,0 +1,146 @@
...
+  const [emailTouched, setEmailTouched] = useState(false);
```
Comment (PiyushMangla) [inline src/Components/Login.jsx]: No need for this

File: src/Components/Login.jsx [hunk 10]
```diff
@@ -0,0 +1,146 @@
...
+    }
+  };
+
+  const handleEmailChange = (e) => {
+    setEmail(e.target.value);
+    if (!emailTouched) setEmailTouched(true);
+    setShowErrorToast(false);
+  };
+
+  const handlePasswordChange = (e) => {
```
Comment (PiyushMangla) [inline src/Components/Login.jsx]: same here

File: src/Components/Login.jsx [hunk 11]
```diff
@@ -0,0 +1,146 @@
...
+            <Form.Label>Password</Form.Label>
+            <div className="position-relative">
+              <Form.Control
+                type={showPassword ? 'text' : 'password'}
+                placeholder="Password"
+                value={password}
+                onChange={handlePasswordChange}
+                onBlur={() => setPasswordTouched(true)}
+                isInvalid={!!passwordError}
+                style={{ paddingRight: '2.5rem' }}
```
Comment (PiyushMangla) [inline src/Components/Login.jsx]: don't use inline styling

File: src/Components/Login.jsx [hunk 12]
```diff
@@ -0,0 +1,146 @@
...
+              />
+              <span
+                className={`position-absolute end-0 top-50 translate-middle-y d-flex justify-content-center align-items-center ${passwordError ? 'mx-5' : 'mx-3'}`}
+                style={{ cursor: 'pointer' }}
```
Comment (PiyushMangla) [inline src/Components/Login.jsx]: here too remove all inline style

File: src/Components/Login.jsx [hunk 13]
```diff
@@ -0,0 +1,146 @@
...
+            type="submit"
+            className="mt-3 w-100 d-flex justify-content-center align-items-center"
+            disabled={isLoading}
+          >
+            {isLoading ? <Spinner animation="border" size="md" /> : 'Login'}
+          </Button>
+        </Form>
+      </Card>
+
+      <ToastContainer
```
Comment (PiyushMangla) [inline src/Components/Login.jsx]: create a component for toast message

File: src/utils/apiClient.jsx [hunk 14]
```diff
@@ -0,0 +1,17 @@
+// src/utils/apiClient.js
+import axios from 'axios';
+
+// Create an Axios instance with a base URL
+const apiClient = axios.create({
+  baseURL: "https://e859-125-63-121-121.ngrok-free.app/", // Your base URL here
```
Comment (PiyushMangla) [inline src/utils/apiClient.jsx]: use env variables if you dont know learn about them

File: .env [hunk 15]
```diff
@@ -0,0 +1,2 @@
+REACT_APP_API_BASE_URL=https://e859-125-63-121-121.ngrok-free.app/
```
Comment (PiyushMangla) [inline .env]: env file should not be pushed. What will be the use of env file if you push it

File: src/modules/Forms/Components/LoginForm.jsx [hunk 16]
```diff
@@ -0,0 +1,154 @@
...
+import { login } from '../../../slices/authSlice';
+import { useNavigate } from 'react-router-dom';
+import { Form, Button, Container, Card, Spinner } from 'react-bootstrap';
+import { EyeFill, EyeSlashFill } from 'react-bootstrap-icons';
+import { isValidEmail } from '../../../utils/validationUtils';
+import ToastMessage from '../Components/ToastMessage';
+
+const Login = () => {
+  const [formData, setFormData] = useState({ email: '', password: '' });
+  const [formFieldTouched, setFormFieldTouched] = useState({
```
Comment (PiyushMangla) [inline src/modules/Forms/Components/LoginForm.jsx]: No need for field touch I already told you

File: src/modules/Forms/Components/LoginForm.jsx [hunk 17]
```diff
@@ -0,0 +1,154 @@
...
+  };
+
+  // Validate password
+  const validatePassword = () => {
+    if (!formFieldTouched.password) return '';
+    if (!formData.password) return 'Password is required';
+    return '';
+  };
+
+  const emailError = validateEmail();
```
Comment (PiyushMangla) [inline src/modules/Forms/Components/LoginForm.jsx]: correct error handling

File: src/modules/Forms/Components/LoginForm.jsx [hunk 18]
```diff
@@ -0,0 +1,154 @@
...
+          password: formData.password.trim()
+        })
+      )
+        .unwrap()
+        .then(() => navigate('/polls'))
+        .catch(() => setShowErrorToast(true));
+    }
+  };
+  // Single handlePasswordEmailChange function for both email and password
+  const handlePasswordEmailChange = (e) => {
```
Comment (PiyushMangla) [inline src/modules/Forms/Components/LoginForm.jsx]: change function name

File: src/Components/TostMassege.jsx [hunk 19]
```diff
@@ -0,0 +1,31 @@
...
+  show,
+  onClose,
+  message,
+  variant = 'danger',
+  delay = 6000
+}) => {
+  return (
+    <ToastContainer
+      position="bottom-center"
+      style={{
```
Comment (yuvrajInno) [inline src/Components/TostMassege.jsx]: remove inline styling

File: src/Pages/LoginForm.jsx [hunk 20]
```diff
@@ -0,0 +1,137 @@
...
+    });
+  };
+
+  const togglePasswordVisibility = () => setShowPassword(!showPassword);
+
+  return (
+    <Container className="d-flex justify-content-center align-items-center vh-100">
+      <Card
+        style={{
+          width: '30rem',
```
Comment (yuvrajInno) [inline src/Pages/LoginForm.jsx]: this too

**Review by yuvrajInno — CHANGES_REQUESTED** [review]

**Pull request merged**